Start optimising the memory footprint of the UTxO.

kantp opened the pull request:
This changes the memory representation of TxIn and TxOut, to reduce overhead.

In particular, we use unpacked Word64 for coin values and txout indices, and use
the serialisation of addresses as ShortByteString.

This does not yet change the representation of TxId.

The changes are nearly opaque, in that we provide view patterns that takes the
place of the old constructors. We do need to add a Crypto constraint wherever we
use a TxOut, however, since the view pattern relies on
serialisation/deserialisation.

This is a draft PR for now, as I'd like to add some properties around the TxOut (I've introduced a `panic` when deserialisation fails in the view pattern, which should never occur, since the constructor itself is not exported, and so the only way to construct a TxOut is using serialisation).

Commit:
Start optimising the memory footprint of the UTxO.

This changes the memory representation of TxIn and TxOut, to reduce overhead.

In particular, we use unpacked Word64 for coin values and txout indices, and use
the serialisation of addresses as ShortByteString.

This does not yet change the representation of TxId.

The changes are nearly opaque, in that we provide view patterns that takes the
place of the old constructors. We do need to add a Crypto constraint wherever we
use a TxOut, however, since the view pattern relies on
serialisation/deserialisation.

dcoutts commented:
Since we rely on the (de)serialiseAddress then we should inherit the round trip property, but it's worth adding a QC test anyway.

Commit:
Run ormolu

Commit:
Adjusting constraints in tests, to make them compile

Commit:
Run ormolu

Commit:
Running ormolu again

Commit:
Properly generate arbitrary UTxO values

Previously, we had relied on Generic, which was using the CompactUTxO
construnctor directly (so the bytestring which was supposed to be a serialised
address was just a random bystestring).

Commit:
Added roundtrip test for TxOut

JaredCorduan reviewed (approved):
LGTM!

dcoutts commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs:
```diff
@@ -375,18 +384,59 @@ deriving newtype instance Crypto crypto => ToCBOR (TxId crypto)
 -- | The input of a UTxO.
-data TxIn crypto
-  = TxIn !(TxId crypto) !Natural -- TODO use our own Natural type
+data TxIn crypto = TxInCompact !(TxId crypto) {-# UNPACK #-} !Word64
```
```suggestion
data TxIn crypto = TxInCompact {-# UNPACK #-} !(TxId crypto) {-# UNPACK #-} !Word64
```
The `TxId` is a single constructor type so we can unpack that one too.

dcoutts commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/TxData.hs:
```diff
@@ -375,18 +384,59 @@ deriving newtype instance Crypto crypto => ToCBOR (TxId crypto)
+pattern TxIn ::
+  Crypto crypto =>
+  TxId crypto ->
+  Natural -> -- TODO We might want to change this to Word64 generally
```
I agree with this TODO.

dcoutts reviewed (approved):
LGTM!

Commit:
unpack the TxId in the TxIn

Co-authored-by: Duncan Coutts <duncan@well-typed.com>
